lps: restore secondary cores on wake up

bkokoszx opened the pull request:
Primary core disables all secondary cores in platform_pg_int_handler() function when going to D0ix state. During coming out from D0ix state (D0ix -> D0), we are not able to restore secondary cores properly. This PR add commits that allow us to enable LPS flow to be used in multicore scenario properly. This PR:
- adds secondary cores data writeback before disabling them;
- adds secondary cores restore during D0->D0ix transition (pm_runtime_get, enable idc interrupt and send power up idc message);
- adds separate booting path in secondary_core_init() function i.e. in case when we have basic firmware structures already allocated (i.e. schedulers, notifier  etc.), which means we are not during cold boot process (memory has not been powered off), we just register and enable required interrupts.

bkokoszx commented on src/init/init.c:
```diff
@@ -70,6 +71,49 @@ static inline void lp_sram_unpack(void)
+	/* initialize interrupts */
+	platform_interrupt_init();
+
+	/* restore schedulers i.e. regist and enable scheduler interrupts */
```
Updated

lyakh commented on src/arch/xtensa/include/arch/lib/cpu.h:
```diff
@@ -28,6 +28,8 @@ int arch_cpu_enabled_cores(void);
 int arch_cpu_restore_secondary_cores(void);
 
+int arch_cpu_writeback_secondary_cores(void);
+
```
to commit description: sorry, I don't see where in `platform_pg_int_handler()` all secondary cores are disabled?

abonislawski replied:
pm_runtime_put

lyakh commented on src/ipc/ipc3/handler.c:
```diff
@@ -705,6 +705,11 @@ static int ipc_pm_gate(uint32_t header)
 		cpu_restore_secondary_cores();
 	} else {
+		/* before we enable pm runtime and perform D0->D0ix flow
+		 * (primary core powers off secondary cores in ) we have to
```
is there a function name missing after "in" above?

lyakh commented on src/ipc/ipc3/handler.c:
```diff
@@ -705,6 +705,11 @@ static int ipc_pm_gate(uint32_t header)
+		 * (primary core powers off secondary cores in ) we have to
+		 * writeback all secondary core data.
+		 */
+		cpu_writeback_secondary_cores();
```
We have flows where the host powers off secondary DSP cores using `SOF_IPC_PM_CORE_ENABLE`. Is `SOF_IPC_PM_GATE` also supposed to be able to switch secondary cores off without `SOF_IPC_PM_CORE_ENABLE` being used? As far as I understand in audio context, `SOF_IPC_PM_GATE` would typically be sent if the only active stream is keyword detection, running on the primary core. In that case all other cores should already be powered off?

abonislawski replied:
If LPS is involved then all secondary cores will be put to sleep by FW independently.
Check issue #4382 for more info about such scenario

abonislawski commented on src/drivers/intel/cavs/idc.c:
```diff
@@ -243,6 +243,35 @@ int platform_idc_init(void)
+	int ret;
+
+	interrupt_disable(idc->irq, idc);
+	interrupt_unregister(idc->irq, idc);
```
Not really, in this scenario core power down sequence is not involved

lgirdwood replied:
ok, so how do we tell the core is is about to power down ? It looks like we enter a WFI and the the LPS HW powers off ?

abonislawski replied:
We wont, lps just powers off - only pm_runtime_put without full power down sequence.
Thats why we have added writeback idc in pm gate handler, secondary core must have a chance to writeback before lps kicks in

abonislawski commented on src/init/init.c:
```diff
@@ -70,6 +71,49 @@ static inline void lp_sram_unpack(void)
+
+	trace_point(TRACE_BOOT_PLATFORM);
+
+	while (1)
```
yes

abonislawski commented on src/schedule/edf_schedule.c:
```diff
@@ -298,6 +298,25 @@ static void scheduler_free_edf(void *data)
+	irq_local_disable(flags);
+
+	interrupt_disable(edf_sch->irq, edf_sch);
+	interrupt_unregister(edf_sch->irq, edf_sch);
```
In this scenario core power down and scheduler_free() will not be called.
There will be nothing to restore if we will call scheduler_free :)

abonislawski commented on src/arch/xtensa/lib/cpu.c:
```diff
@@ -189,3 +189,28 @@ void cpu_power_down_core(void)
 		arch_wait_for_interrupt(0);
 }
+
+int arch_cpu_restore_secondary_cores(void)
```
This code will only restore cores which was active previously, such a mask will be used inside this function (it is not available globally)

lgirdwood replied:
Why can we not have a global ```SHARED``` flag, possible as part of sof context ? I'm assuming this would be valid since it would be in the uncache alias.

abonislawski replied:
No cache issues here (only core0 will access it). That flag already exists - active_cores_mask, its static in cpu.c (but can be used anywhere from cpu_enabled_cores() if needed), 
It is used inside arch_cpu_restore_secondary_cores() (in arch_cpu_is_core_enabled(id) checks).

We can take it from cpu.c as cpu_enabled_cores(), run cpu_restore_secondary_cores(cpu_enabled_cores()) and inside anyway run helper functions like arch_cpu_is_core_enabled() which have local access to active cores mask. All of this is just weird and unnecessary. Maybe the name is just misleading and it will be better to call it cpu_restore_active_secondary_cores or something?

abonislawski commented on src/init/init.c:
```diff
@@ -70,6 +71,49 @@ static inline void lp_sram_unpack(void)
+	/* check whether basic core structures has been already allocated. If they
+	   are available in memory, it means that this is not cold boot and memory
+	   has not been powered off. */
+	if (!idc || !task || !notifier || !schedulers)
```
Good idea but I think we should still somehow validate all critical values in memory to make sure it is still valid to restore

lgirdwood commented on src/platform/intel/cavs/lib/pm_runtime.c:
```diff
@@ -567,6 +570,30 @@ void platform_pm_runtime_enable(uint32_t context, uint32_t index)
+	struct pm_runtime_data *prd = pm_runtime_data_get();
+	struct cavs_pm_runtime_data *pprd = prd->platform_data;
+
+	pprd->wait_writeback_core_mask |= BIT(index);
```
Maybe I've mentioned this in another patch, but dont we need to lock this mask ?

abonislawski replied:
I think we should

keyonjie commented:
> @bkokoszx @keyonjie @ranj063 lets have a proper shutdown sequence, this PR should not have to shutdown and reinitialize subsystems in the bringup. Topic for Tuesday call.

yes, we need to do more tearing down before shutting down a secondary core, I drafted a PR to showcase what I meant in the call here:
https://github.com/thesofproject/sof/pull/4718

bkokoszx commented:
@lgirdwood @keyonjie @abonislawski @lyakh 
I've just updated this PR. I've added possibility for idc_free() and schedule_free() to disable only interrupts. I've also added flag for ```cpu_power_down_core()``` function i.e. CPU_POWER_DOWN_MEMORY_ON, which indicates not to free memory resources (basically it disables interrupts and perform writeback/invalidate of data) - ```cpu_power_down_core(CPU_POWER_DOWN_MEMORY_ON)``` is invoked for secondary cores in ```platform_wait_for_interrupt()``` when we should make a preparation before going to d0ix state.

To sum up:
- before going to d0ix state secondary cores performs writeback and disables interrupts;
- during booting up (after d0ix) secondary cores just enable idc and scheduler interrupts.

lgirdwood commented:
@bkokoszx  the DMIC CI failures look unrelated for this PM related PR ?

bkokoszx commented:
@lgirdwood 
I am not sure whether TIMEOUT on CML_RVP_SDW in ``check-suspend-resume-with-playback-5.sh`` test case is releated. If test just pauses and resumes pipelines without any D0ix flow then it should not be related with above change.

BTW
I have made one little, additional change:

``` c
static int ipc_pm_gate(uint32_t header)
{
	struct sof_ipc_pm_gate pm_gate;

	IPC_COPY_CMD(pm_gate, ipc_get()->comp_data);

	tr_info(&ipc_tr, "ipc: pm gate flags 0x%x", pm_gate.flags);

	/* pause dma trace firstly if needed */
	if (pm_gate.flags & SOF_PM_NO_TRACE)
		trace_off();

	if (pm_gate.flags & SOF_PM_PPG) {
		pm_runtime_disable(PM_RUNTIME_DSP, PLATFORM_PRIMARY_CORE_ID);
#if (CONFIG_CAVS_LPS)
		cpu_restore_secondary_cores();
#endif
	} else {
		/* before we enable pm runtime and perform D0->D0ix flow
		 * (primary core powers off secondary cores in
		 * platform_pg_int_handler) we have to prepare all secondary
		 * cores data for powering off (disable interrupt, perform
		 * cache writeback).
		 */
#if (CONFIG_CAVS_LPS)
		cpu_secondary_cores_prepare_d0ix();
#endif
		pm_runtime_enable(PM_RUNTIME_DSP, PLATFORM_PRIMARY_CORE_ID);
	}

	/* resume dma trace if needed */
	if (!(pm_gate.flags & SOF_PM_NO_TRACE))
		trace_on();

	return 0;
}
```
I invoke ```cpu_restore_secondary_cores()``` and ```cpu_secondary_cores_prepare_d0ix()``` only in case when LPS is enabled.

bkokoszx commented:
@lgirdwood  Do you require some other changes?
